### PR TITLE
Reagent tank optimization

### PR DESCRIPTION
**Pull request by ghost**
- Reagent dispensers now use two different variables - `vol` and `reagent_id` - to determine what reagent to dispense and how much of it, instead of using a `New()` overwrite.
- Subtypes of `/obj/structure/reagent_dispenser` no longer redefine the `icon` variable if it's the same as the parent type.
- Fuel tanks now have an `attackby()` for refueling welding tools instead of an `afterattack()` for the welding tools themselves.
- High-capacity water tanks now have the intended amount of water.
- Removed snowflake checks for water coolers (excuse me, _liquid_ coolers).
- Fuel tanks no longer explode if they're empty.
- Tweaked the names and descriptions of various subtypes.

### CODE REVIEW

**Comment by Jordie0608** [comment]
> Fuel tanks no longer explode if they're empty

Empty welding tanks are still dangerous due to residue, however this could easily be a gameplay abstraction

**Comment by Xhuis** [comment]
Currently, the tanks themselves explode violently with a huge fireball regardless of how much fuel is inside of them. I don't know how much damage residue could cause realistically, but if you know about this kind of thing, I'd be happy to remove the change.

**Comment by MMMiracles** [comment]
>removing the ability to get paper cups from the water cooler

stealth nerf, close PR immediately

**Comment by Xhuis** [comment]
Oh, I didn't read that part fully and just assumed that it was snowflake checking. I'll put that back when I can.

**Comment by Jordie0608** [comment]
I googled this possibly dubious video https://www.youtube.com/watch?v=FK-S0JdE_6U, although that's petrol and not welding fuel (acetylene) I don't think it's worth discussing if/how large a fireball would be produced. Just go for what works best in gameplay.

**Comment by coiax** [comment]
I've never seen an empty welding fuel tank.

I've emptied a water tank via atmos firepack, that can be done pretty easily with nanofrost spam.

File: code/modules/reagents/reagent_dispenser.dm [hunk 1]
```diff
@@ -141,40 +157,18 @@
 	icon = 'icons/obj/vending.dmi'
 	icon_state = "water_cooler"
 	anchored = 1
-	var/cups = 50
```
Comment (KazeEspada) [inline code/modules/reagents/reagent_dispenser.dm]: Nooooooooooooooooo muh silly cups.

File: code/modules/reagents/reagent_dispenser.dm [hunk 2]
```diff
@@ -122,17 +114,41 @@
...
+			playsound(src, 'sound/effects/refill.ogg', 50, 1)
+			update_icon()
+			return
+		else
+			user.visible_message("<span class='warning'>[user] detonates [src]!</span>", "<span class='userdanger'>That was stupid of you.</span>")
+			message_admins("[key_name_admin(user)] triggered a fueltank explosion.")
+			log_game("[key_name(user)] triggered a fueltank explosion.")
+			boom()
+			return
+
```
Comment (phil235) [inline code/modules/reagents/reagent_dispenser.dm]: else 
	return ..()

File: code/modules/reagents/reagent_dispenser.dm [hunk 3]
```diff
@@ -2,10 +2,12 @@
 	name = "Dispenser"
 	desc = "..."
 	icon = 'icons/obj/objects.dmi'
-	icon_state = "watertank"
+	icon_state = "water"
 	density = 1
 	anchored = 0
 	pressure_resistance = 2*ONE_ATMOSPHERE
+	var/vol = 1000 //In units, how much the dispenser can hold
```
Comment (phil235) [inline code/modules/reagents/reagent_dispenser.dm]: a bit too short for a var name imo. Maybe "tank_volume" ?

File: code/modules/reagents/reagent_dispenser.dm [hunk 4]
```diff
@@ -185,10 +179,9 @@
 
 
 /obj/structure/reagent_dispensers/virusfood
-	name = "Virus Food Dispenser"
-	desc = "A dispenser of virus food."
-	icon = 'icons/obj/objects.dmi'
-	icon_state = "virusfoodtank"
+	name = "virus food dispenser"
+	desc = "A dispenser of low-potency virus mutagenic."
+	icon_state = "virus_food"
```
Comment (phil235) [inline code/modules/reagents/reagent_dispenser.dm]: no reagent_id = "virusfood" ?

Comment (Xhuis) [inline code/modules/reagents/reagent_dispenser.dm]: Whoops, forgot about that. I'll address this and your other concerns once I can.

**Comment by coiax** [comment]
I did some of these changes during my bucket buff PR, it's good to see them.

**Comment by Xhuis** [comment]
_New in this commit:_
* Re-added paper cups to water coolers.
* `vol` is now `tank_volume`.
* Admin logging now shows how a fuel tank was detonated.
* Tweaked some examine texts and messages.
* Removed two `New()` overrides in favor of the new system.

**Comment by optimumtact** [comment]
@Xhuis this seems to have broken welding tank refueling

**Comment by optimumtact** [comment]
#18751

File: code/modules/reagents/reagent_dispenser.dm [hunk 5]
```diff
@@ -122,17 +114,41 @@
...
+	if(istype(I, /obj/item/weapon/weldingtool))
+		if(!reagents.has_reagent("welding_fuel"))
+			user << "<span class='warning'>[src] is out of fuel!</span>"
+			return
+		var/obj/item/weapon/weldingtool/W = I
+		if(!W.welding)
+			if(W.reagents.has_reagent("welding_fuel", W.max_fuel))
+				user << "<span class='warning'>Your [W.name] is already full!</span>"
+				return
+			reagents.trans_to(src, W.max_fuel)
```
Comment (phil235) [inline code/modules/reagents/reagent_dispenser.dm]: reagents.trans_to(<b>W</b>, W.max_fuel)

**Pull request merged**